\chapter{Hybrid Optical Map-\/\+Guided Assembly}
\hypertarget{md_docs_2genome-assembly-guide}{}\label{md_docs_2genome-assembly-guide}\index{Hybrid Optical Map-\/Guided Assembly@{Hybrid Optical Map-\/Guided Assembly}}
\label{md_docs_2genome-assembly-guide_autotoc_md0}%
\Hypertarget{md_docs_2genome-assembly-guide_autotoc_md0}%
 \hypertarget{md_docs_2genome-assembly-guide_autotoc_md1}{}\doxysection{\texorpdfstring{Download Test Data}{Download Test Data}}\label{md_docs_2genome-assembly-guide_autotoc_md1}
Input data is from GIAB\+: Oxford Nanopore ultralong (guppy-\/\+V3.\+2.\+4\+\_\+2020-\/01-\/22)

\href{https://github.com/genome-in-a-bottle/giab_data_indexes}{\texttt{ https\+://github.\+com/genome-\/in-\/a-\/bottle/giab\+\_\+data\+\_\+indexes}}\hypertarget{md_docs_2genome-assembly-guide_autotoc_md2}{}\doxysection{\texorpdfstring{De novo Assembly}{De novo Assembly}}\label{md_docs_2genome-assembly-guide_autotoc_md2}
We will perform de novo assembly of the hg002 data using miniasm (instructions from \href{https://github.com/lh3/miniasm}{\texttt{ https\+://github.\+com/lh3/miniasm}})


\begin{DoxyItemize}
\item Also see Optical Kermit\textquotesingle{}s optical map guided assembly script, which uses the same tools and approach (\href{https://github.com/Denopia/kermit-optical-maps/blob/master/genome-assembly.sh}{\texttt{ https\+://github.\+com/\+Denopia/kermit-\/optical-\/maps/blob/master/genome-\/assembly.\+sh}})
\end{DoxyItemize}\hypertarget{md_docs_2genome-assembly-guide_autotoc_md3}{}\doxysubsection{\texorpdfstring{Set up the software environment\+:}{Set up the software environment:}}\label{md_docs_2genome-assembly-guide_autotoc_md3}

\begin{DoxyCode}{0}
\DoxyCodeLine{export\ PATH="{}\$software\_dir"{}/minimap2:\$PATH}
\DoxyCodeLine{export\ PATH="{}\$software\_dir"{}/miniasm:\$PATH}

\end{DoxyCode}
\hypertarget{md_docs_2genome-assembly-guide_autotoc_md4}{}\doxysubsection{\texorpdfstring{Use minimap2 to self-\/align the hg002 data\+:}{Use minimap2 to self-align the hg002 data:}}\label{md_docs_2genome-assembly-guide_autotoc_md4}

\begin{DoxyCode}{0}
\DoxyCodeLine{\ "{}Running\ minimap2..."{}}
\DoxyCodeLine{minimap2\ -\/x\ ava-\/ont\ -\/t\ 12\ input\_data/HG002\_hs37d5\_ONT\_GIAB.fastq\ input\_data/HG002\_hs37d5\_ONT\_GIAB.fastq\ |\ gzip\ -\/1\ >\ output/HG002\_hs37d5\_ONT\_GIAB.paf.gz}

\end{DoxyCode}
\hypertarget{md_docs_2genome-assembly-guide_autotoc_md5}{}\doxysubsection{\texorpdfstring{Use miniasm to assemble the self-\/aligned data\+:}{Use miniasm to assemble the self-aligned data:}}\label{md_docs_2genome-assembly-guide_autotoc_md5}

\begin{DoxyCode}{0}
\DoxyCodeLine{echo\ "{}Running\ miniasm..."{}}
\DoxyCodeLine{miniasm\ -\/f\ inputs/HG002\_hs37d5\_ONT\_GIAB.fastq\ outputs/HG002\_hs37d5\_ONT\_GIAB.paf.gz\ >\ outputs/HG002\_hs37d5\_ONT\_GIAB.gfa}

\end{DoxyCode}
\hypertarget{md_docs_2genome-assembly-guide_autotoc_md6}{}\doxysubsection{\texorpdfstring{Convert GFA file to FASTA for downstream analysis}{Convert GFA file to FASTA for downstream analysis}}\label{md_docs_2genome-assembly-guide_autotoc_md6}
(instructions from \href{https://www.biostars.org/p/169516/}{\texttt{ https\+://www.\+biostars.\+org/p/169516/}})


\begin{DoxyCode}{0}
\DoxyCodeLine{echo\ "{}Converting\ GFA\ to\ FASTA..."{}}
\DoxyCodeLine{awk\ '/\string^S/\{print\ "{}>"{}\$2"{}\(\backslash\)n"{}\$3\}'\ outputs/HG002\_hs37d5\_ONT\_GIAB.gfa\ >\ outputs/HG002\_hs37d5\_ONT\_GIAB.fasta}
\DoxyCodeLine{echo\ "{}Complete."{}}

\end{DoxyCode}


This yields an assembly which can be used for input into Bio\+Nano\textquotesingle{}s hybrid scaffolding pipeline to generate a more accurate hybrid assembly using the Bio\+Nano data. 